rand_distr/src/dirichlet.rs

```diff
@@ -9,11 +9,160 @@
+#[cfg_attr(feature = "serde1", derive(serde::Serialize, serde::Deserialize))]
+struct DirichletFromGamma<F>
+where
+    F: Float,
+    StandardNormal: Distribution<F>,
+    Exp1: Distribution<F>,
+    Open01: Distribution<F>,
+{
+    samplers: Box<[Gamma<F>]>,
+}
+
+impl<F> DirichletFromGamma<F>
+where
+    F: Float,
+    StandardNormal: Distribution<F>,
+    Exp1: Distribution<F>,
+    Open01: Distribution<F>,
+{
+    // Construct a new `Dirichlet` with the given alpha parameter `alpha`.
+    //
+    // This function is part of a private implementation detail.
+    // It assumes that the input is correct, so no validation is done.
+    #[inline]
+    fn new(alpha: &[F]) -> DirichletFromGamma<F> {
+        let gamma_dists = alpha
+            .iter()
+            .map(|a| Gamma::new(*a, F::one()).unwrap())
+            .collect::<Vec<Gamma<F>>>()
+            .into_boxed_slice();
+        DirichletFromGamma {
+            samplers: gamma_dists,
+        }
+    }
+}
+
+impl<F> Distribution<Vec<F>> for DirichletFromGamma<F>
+where
+    F: Float,
+    StandardNormal: Distribution<F>,
+    Exp1: Distribution<F>,
+    Open01: Distribution<F>,
+{
+    fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> Vec<F> {
+        let n = self.samplers.len();
+        let mut samples = vec![F::zero(); n];
+        let mut sum = F::zero();
+        for (s, g) in samples.iter_mut().zip(self.samplers.iter()) {
+            *s = g.sample(rng);
+            sum = sum + (*s);
+        }
+        let invacc = F::one() / sum;
+        for s in samples.iter_mut() {
+            *s = (*s) * invacc;
+        }
+        samples
+    }
+}
+
+#[derive(Clone, Debug, PartialEq)]
+#[cfg_attr(feature = "serde1", derive(serde::Serialize, serde::Deserialize))]
+struct DirichletFromBeta<F>
+where
+    F: Float,
+    StandardNormal: Distribution<F>,
+    Exp1: Distribution<F>,
+    Open01: Distribution<F>,
+{
+    samplers: Box<[Beta<F>]>,
+}
+
+impl<F> DirichletFromBeta<F>
+where
+    F: Float,
+    StandardNormal: Distribution<F>,
+    Exp1: Distribution<F>,
+    Open01: Distribution<F>,
+{
+    // Construct a new `Dirichlet` with the given alpha parameter `alpha`.
+    //
+    // This function is part of a private implementation detail.
+    // It assumes that the input is correct, so no validation is done.
+    #[inline]
+    fn new(alpha: &[F]) -> DirichletFromBeta<F> {
+        // Form the right-to-left cumulative sum of alpha, exluding the
+        // first element of alpha.  E.g. if alpha = [a0, a1, a2, a3], then
+        // after the call to `alpha_sum_rl.reverse()` below, alpha_sum_rl
+        // will hold [a1+a2+a3, a2+a3, a3].
+        let mut alpha_sum_rl: Vec<F> = alpha
+            .iter()
+            .skip(1)
+            .rev()
+            // scan does the cumulative sum
+            .scan(F::zero(), |sum, x| {
+                *sum = *sum + *x;
+                Some(*sum)
+            })
+            .collect();
+        alpha_sum_rl.reverse();
```

Comment:
This looks like we could avoid one pass over `alpha` by assigning the values in a loop in the correct order.

Reply:
Done.